File: impl/go/internal/tool/registry.go
// Tool registry (Go parity of registry.ts). See docs/harness-design.md §2, §4.
// Holds id → Def, validates args, runs Execute through a shared wrapper: a
// bad-args call becomes an ok:false result (model is asked to rewrite) instead
// of erroring; panics/errors are caught so one failing tool never crashes the
// loop; every call can log a transcript step.
package tool

import (
	"fmt"
	"strings"
)

// builtin tools, in catalog order. Domain tools are appended via NewRegistry's
// extra arg (the projects package supplies them).
func builtinTools() []Def {
	return []Def{ReadTool, WriteTool, EditTool, ListFilesTool, BashTool, GrepTool, GlobTool, RenderSceneTool, ReadSkillRuleTool}
}

type Registry struct {
	tools map[string]Def
	order []string
}

func NewRegistry(extra ...Def) *Registry {
	r := &Registry{tools: map[string]Def{}}
	for _, t := range append(builtinTools(), extra...) {
		if _, ok := r.tools[t.ID]; !ok {
			r.order = append(r.order, t.ID)
		}
		r.tools[t.ID] = t
	}
	return r
}

func (r *Registry) IDs() []string { return append([]string(nil), r.order...) }

func (r *Registry) IsMutating(id string) bool {
	t, ok := r.tools[id]
	return ok && t.Mutating
}

// Execute runs one call through the shared wrapper. Never returns an error —
// always a CallResult the director can render back to the model.
func (r *Registry) Execute(call Call, ctx *Context) CallResult {
	def, ok := r.tools[call.Tool]
	if !ok {
		res, _ := InvalidTool.Execute(map[string]any{"tool": call.Tool, "error": fmt.Sprintf("unknown tool %q", call.Tool)}, ctx)
		return CallResult{Tool: call.Tool, OK: false, Output: res.Output, Title: res.Title}
	}

	args := call.Args
	if args == nil {
		args = map[string]any{}
	}
	validated, err := def.Validate(args)
	if err != nil {
		ctx.LogStep("tool", fmt.Sprintf("%s: invalid args — %s", call.Tool, err.Error()), StepExtra{Kind: "error"})
		return CallResult{
			Tool:   call.Tool,
			OK:     false,
			Output: fmt.Sprintf("The %s tool was called with invalid arguments: %s\nRewrite the input so it satisfies the parameters.", call.Tool, err.Error()),
		}
	}

	res, execErr := safeExecute(def, validated, ctx)
	if execErr != nil {
		ctx.LogStep("tool", fmt.Sprintf("%s: failed — %s", call.Tool, execErr.Error()), StepExtra{Kind: "error"})
		return CallResult{Tool: call.Tool, OK: false, Output: fmt.Sprintf("Tool %s failed: %s", call.Tool, execErr.Error())}
	}
	// A tool may report a soft failure via metadata.error — surface as ok:false.
	_, softFail := res.Metadata["error"]
	return CallResult{
		Tool:       call.Tool,
		OK:         !softFail,
		Output:     res.Output,
		Title:      res.Title,
		Metadata:   res.Metadata,
		Truncated:  res.Truncated,
		OutputPath: res.OutputPath,
	}
}

// safeExecute runs a tool's Execute, converting a panic into an error so one
// misbehaving tool can't take down the agent loop.
func safeExecute(def Def, args map[string]any, ctx *Context) (res Result, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("panic: %v", r)
		}
	}()
	return def.Execute(args, ctx)
}

// RenderDocs renders the tool catalog for the system prompt.
func (r *Registry) RenderDocs() string {
	var b strings.Builder
	for _, id := range r.order {
		t := r.tools[id]
		fmt.Fprintf(&b, "- %s: %s", t.ID, t.Description)
		for _, p := range t.Parameters {
			req := ""
			if p.Required {
				req = " (required)"
			}
			items := ""
			if p.Items != "" {
				items = " of " + p.Items
			}
			fmt.Fprintf(&b, "\n    - %s: %s%s%s — %s", p.Name, p.Type, items, req, p.Description)
		}
		b.WriteString("\n")
	}
	return strings.TrimRight(b.String(), "\n")
}

// AppendDiagnostics runs one project type-check (if any mutating call succeeded)
// and splices the errors into the matching mutating results — see §8. Mutates
// results in place.
func (r *Registry) AppendDiagnostics(results []CallResult, ctx *Context) {
	anyMutated := false
	for _, res := range results {
		if res.OK && r.IsMutating(res.Tool) {
			anyMutated = true
			break
		}
	}
	if !anyMutated {
		return
	}
	byFile := typecheck(ctx)
	if len(byFile) == 0 {
		return
	}
	for i := range results {
		if !(results[i].OK && r.IsMutating(results[i].Tool)) {
			continue
		}
		edited, _ := results[i].Metadata["filePath"].(string)
		if text := formatDiagnostics(byFile, edited); text != "" {
			results[i].Output += text
		}
	}
}

File: impl/ts/src/tool/registry.ts
// Tool registry — see docs/harness-design.md §2, §4.
// Holds id → ToolDef, validates args, runs execute() through a shared wrapper
// that mirrors opencode's tool.ts: a bad-args call becomes an ok:false result
// (model is asked to rewrite) instead of throwing; exceptions are caught so one
// failing tool never crashes the agent loop; every call logs a transcript step.

import type { ToolCall, ToolCallResult, ToolContext, ToolDef } from "./types.js";
import { ToolValidationError } from "./types.js";
import { ReadTool } from "./read.js";
import { WriteTool } from "./write.js";
import { EditTool } from "./edit.js";
import { ListFilesTool } from "./list-files.js";
import { BashTool } from "./bash.js";
import { GrepTool } from "./grep.js";
import { GlobTool } from "./glob.js";
import { InvalidTool } from "./invalid.js";
import { typecheck, formatDiagnostics } from "./diagnostics.js";

// Ordered list of built-in tools. Increment D appends domain tools.
const BUILTIN: ToolDef<any>[] = [ReadTool, WriteTool, EditTool, ListFilesTool, BashTool, GrepTool, GlobTool];

export class ToolRegistry {
  private tools = new Map<string, ToolDef<any>>();

  constructor(extra: ToolDef<any>[] = []) {
    for (const t of [...BUILTIN, ...extra]) this.tools.set(t.id, t);
  }

  ids(): string[] {
    return [...this.tools.keys()];
  }

  // Run one tool call through the shared wrapper. Never throws — always returns
  // a ToolCallResult the director can render back to the model.
  async execute(call: ToolCall, ctx: ToolContext): Promise<ToolCallResult> {
    const def = this.tools.get(call.tool);
    if (!def) {
      const r = await InvalidTool.execute(
        { tool: call.tool, error: `unknown tool "${call.tool}"` },
        ctx,
      );
      return { tool: call.tool, ok: false, output: r.output, title: r.title };
    }

    let validated: unknown;
    try {
      validated = def.validate(call.args ?? {});
    } catch (e) {
      const detail = e instanceof ToolValidationError ? e.message : String((e as Error).message ?? e);
      ctx.log("tool", `${call.tool}: invalid args — ${detail}`, { kind: "error" });
      return {
        tool: call.tool,
        ok: false,
        output: `The ${call.tool} tool was called with invalid arguments: ${detail}\nRewrite the input so it satisfies the parameters.`,
      };
    }

    try {
      const result = await def.execute(validated, ctx);
      // A tool may report a soft failure (operation didn't apply) by setting
      // metadata.error instead of throwing — surface it as ok:false so the
      // model knows to retry, while still showing the explanatory output.
      const softFail = typeof result.metadata?.error === "string";
      return {
        tool: call.tool,
        ok: !softFail,
        output: result.output,
        title: result.title,
        metadata: result.metadata,
        truncated: result.truncated,
        outputPath: result.outputPath,
      };
    } catch (e) {
      const detail = String((e as Error).message ?? e);
      ctx.log("tool", `${call.tool}: failed — ${detail}`, { kind: "error" });
      return { tool: call.tool, ok: false, output: `Tool ${call.tool} failed: ${detail}` };
    }
  }

  // True if the named tool mutates files (write/edit). The director uses this to
  // decide whether a diagnostics pass is worth running after a turn.
  isMutating(toolId: string): boolean {
    return this.tools.get(toolId)?.mutating === true;
  }

  // Run one project type-check and append the errors to the matching mutating
  // tool results, so the model sees compile errors in the same turn it edited —
  // see docs/harness-design.md §8. Runs at most once per turn (tsc is costly),
  // only when a mutating call actually succeeded. Mutates `results` in place.
  async appendDiagnostics(results: ToolCallResult[], ctx: ToolContext): Promise<void> {
    const mutated = results.filter((r) => r.ok && this.isMutating(r.tool));
    if (mutated.length === 0) return;
    const byFile = await typecheck(ctx);
    if (byFile.size === 0) return;
    for (const r of mutated) {
      const edited = typeof r.metadata?.filePath === "string" ? r.metadata.filePath : "";
      const text = formatDiagnostics(byFile, edited);
      if (text) r.output += text;
    }
  }

  // Render the tool catalog for the system prompt: id, description, params.
  renderDocs(): string {
    const blocks: string[] = [];
    for (const t of this.tools.values()) {
      const params = Object.entries(t.parameters)
        .map(([name, spec]) => {
          const req = spec.required ? " (required)" : "";
          const items = spec.items ? ` of ${spec.items}` : "";
          return `    - ${name}: ${spec.type}${items}${req} — ${spec.description}`;
        })
        .join("\n");
      blocks.push(`- ${t.id}: ${t.description}${params ? "\n" + params : ""}`);
    }
    return blocks.join("\n");
  }
}
